src/main/haskell/language-kore/src/Data/Kore/ASTUtils/Substitution.hs

```diff
@@ -0,0 +1,124 @@
+{-|
+Module      : Data.Kore.ASTUtils.Substitution
+Description : Substitute phi_1 for phi_2, avoiding capture
+Copyright   : (c) Runtime Verification, 2018
+License     : UIUC/NCSA
+Maintainer  : phillip.harris@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE Rank2Types             #-}
+{-# LANGUAGE AllowAmbiguousTypes    #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE BangPatterns           #-}
+{-# LANGUAGE LambdaCase             #-}
+{-# LANGUAGE DeriveFunctor          #-}
+{-# LANGUAGE DeriveFoldable         #-}
+{-# LANGUAGE DeriveTraversable      #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE PatternSynonyms        #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TypeApplications       #-}
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+module Data.Kore.ASTUtils.Substitution where
+
+
+import           Control.Lens
+
+import qualified Data.Set as S
+import           Data.Fix
+import           Data.List
+
+
+import           Data.Kore.AST.Common
+import           Data.Kore.AST.MetaOrObject
+import           Data.Kore.AST.PureML
+
+import           Data.Kore.ASTUtils.SmartConstructors
+
+
+subst
+  :: MetaOrObject level 
+  => CommonPurePattern level 
+  -> CommonPurePattern level 
+  -> CommonPurePattern level 
+  -> CommonPurePattern level
+subst a b (Forall_ s1 v p) = handleBinder a b Forall_ s1 v p
+subst a b (Exists_ s1 v p) = handleBinder a b Exists_ s1 v p
+subst a b pat =
+  if pat == a then b else Fix $ fmap (subst a b) $ unFix pat
```

Comment:
I think `subst` could actually be rewritten as a catamorphism using `embed`. Maybe that's a little tricky if we're doing whole-term substitution instead of only variables. Although, maybe not: I don't see that `pat` could contain `a` as a subterm and also be equal to `a`.